pkg/virt-handler/migration-target.go

```diff
@@ -722,36 +737,39 @@ func (c *MigrationTargetController) processVMI(vmi *v1.VirtualMachineInstance) e
 	}
 
 	if err := client.SyncMigrationTarget(vmi, options); err != nil {
-		return fmt.Errorf("syncing migration target failed: %v", err)
+		return fmt.Errorf("syncing migration target failed: %v", err), false
 	}
 
 	err = c.handleTargetMigrationProxy(vmi)
 	if err != nil {
-		return fmt.Errorf("failed to handle post sync migration proxy: %v", err)
+		return fmt.Errorf("failed to handle post sync migration proxy: %v", err), false
 	}
 
 	c.recorder.Event(vmi, k8sv1.EventTypeNormal, v1.PreparingTarget.String(), VMIMigrationTargetPrepared)
 
-	return nil
+	return nil, false
 }
 
 func (c *MigrationTargetController) addFunc(obj interface{}) {
 	key, err := controller.KeyFunc(obj)
 	if err == nil {
+		c.vmiExpectations.SetExpectations(key, 0, 0)
 		c.queue.Add(key)
 	}
 }
 
 func (c *MigrationTargetController) deleteFunc(obj interface{}) {
 	key, err := controller.KeyFunc(obj)
 	if err == nil {
+		c.vmiExpectations.SetExpectations(key, 0, 0)
 		c.queue.Add(key)
 	}
 }
 
 func (c *MigrationTargetController) updateFunc(_, new interface{}) {
 	key, err := controller.KeyFunc(new)
 	if err == nil {
+		c.vmiExpectations.SetExpectations(key, 0, 0)
```

Comment:
I think this is the cause of the deadlock. I don't think it's needed.

Reply:
This is what clears the expectation, I'm not sure how it could cause a deadlock...
The mechanism is:
- Whenever we successfully update the VMI we set (add) expectation to 1
- Whenever the informer registers that update we reset it to 0 (here)
- If, between the two, we enter an execution cycle, we will bail because the add expectation is 1

Am I still missing something? (hopefully :))

Reply:
Ok, this is very interesting... To me, expectations were just a pair of ints that you can atomically inc/dec/set, but it looks like it's more than that.
I thought `SatisfyExpectations()` was the same as `SetExpectations(0)`, good to know it's not!
However, about the `vmiExpectations` `UIDTrackingControllerExpectations`, how does it even know it is tracking VMIs? Nothing in the the variable definition says so... Does it get auto-satisfied whenever any object that matches the namespace+name of the key gets updated?
Thanks

Reply:
We can print the values e.add.Load() ,  e.del.Load() to see what's there, but also
How do we know on `update` which controller satisfies the expectations of who?

Reply:
> We can print the values e.add.Load() , e.del.Load() to see what's there, but also How do we know on `update` which controller satisfies the expectations of who?

Each controller maintains its own vmiExpectations variable.

Reply:
We actually have one informer per controller too, not that it makes much of a difference though.
I am under the (potentially erroneous) assumption that if you see an informer update any time after a successful API call to `Update()`, then the VMI in the informer is now guaranteed to include whatever was updated by the API call. If that's incorrect, we do indeed have a problem, in every controller that uses expectations, and it might be impossible to solve with our current tools.
However, that problem wouldn't lead to deadlocks, quite the opposite (more informer update calls = more expectation resets = lower chances of tripping on `SatisfiedExpectations`).